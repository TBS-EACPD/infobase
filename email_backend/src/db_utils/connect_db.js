import mongoose from "mongoose";

/* eslint-disable no-console */

const get_connection_str = () => {
  if(process.env.USE_REMOTE_DB){
    console.log("using remote db")
    const {
      MDB_CONNECT_STRING,
      MDB_USERNAME,
      MDB_PW,
      MDB_NAME,
    } = process.env;

    return MDB_CONNECT_STRING
      .replace('{MDB_USERNAME}', MDB_USERNAME)
      .replace('{MDB_PW}', MDB_PW)
      .replace('{MDB_NAME}', MDB_NAME);
  } else {
    console.log("using local db")
    return "mongodb://127.0.0.1:27018/email_backend";
  }  
};


export const connect_db = async () => {
  return await mongoose.connect(
    get_connection_str(),
    { 
      useCreateIndex: true,
      useNewUrlParser: true,
      poolSize: 10,
    }
  )
    .then( () => console.log("MongoDB connected") )
    .catch( console.log );
};