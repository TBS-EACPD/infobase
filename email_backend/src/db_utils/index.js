export { connect_db } from './connect_db.js'
