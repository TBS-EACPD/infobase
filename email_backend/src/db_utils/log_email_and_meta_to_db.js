import mongoose from "mongoose";
import _ from "lodash";

const required_string = { type: String, required: true };
const meta_schema_def = {
  to: required_string,
  from: required_string,
  referer: required_string,
  server_time: { type: Date, required: true },
  date: required_string,
  time: required_string,
};
const get_meta_fields_for_log = (email_config, request) => {
  const { to, from } = email_config;
  const {
    headers: { referer },
  } = request;

  const server_time = new Date();

  return {
    to,
    from,
    referer,
    server_time,
    date: server_time.toLocaleDateString("en-CA"),
    time: server_time.toLocaleTimeString("en-CA"),
  };
};

const template_type_to_schema_type = (value_type) => {
  switch (value_type) {
    case "enums":
      return [String];
    case "json":
      return Object;
    case "number":
      return Number;
    default:
      return String;
  }
};
export const make_mongoose_model_from_original_template = _.memoize(
  ({ original_template, template_name }) => {
    const template_schema = _.chain(original_template)
      .mapValues(({ required, value_type }) => ({
        required,
        type: template_type_to_schema_type(value_type),
      }))
      .thru((template_schema_def) =>
        mongoose.Schema({
          ...template_schema_def,
          email_meta: meta_schema_def,
        })
      )
      .value();

    const model_name = `${template_name}_emails`;

    // Creating a mongoose model has the side effect of initializing a corresponding collection on the
    // connected DB. This operation will buffer if the DB is not yet connected
    return mongoose.model(model_name, template_schema);
  },
  ({ template_name }) => template_name
);

export const log_email_and_meta_to_db = async (
  request,
  template_name,
  original_template,
  completed_template,
  email_config
) => {
  const model = make_mongoose_model_from_original_template({
    template_name,
    original_template,
  });

  const meta_sub_doc = get_meta_fields_for_log(email_config, request);

  return model.create({
    ...completed_template,
    email_meta: meta_sub_doc,
  });
};
