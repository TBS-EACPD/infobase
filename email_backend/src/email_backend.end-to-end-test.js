import axios from 'axios';
import _ from 'lodash';


const promise_timeout_race = (promise, time_limit, timeout_callback=_.noop) => {
  const timeout_promise = new Promise((resolve, reject) => {
    const timeout = setTimeout(
      () => {
        clearTimeout(timeout);

        timeout_callback(resolve, reject)
      },
      time_limit
    );
  });

  return Promise.race([promise, timeout_promise]);
};


// mocking JUST nodemailer.createTransport, leaving everything else with its original implementation... is there a more direct way to do this?
jest.mock('nodemailer'); // eslint-disable-line no-undef
import nodemailer from 'nodemailer';
const { ...actual_nodemailer } = jest.requireActual('nodemailer'); // eslint-disable-line no-undef

_.each(
  nodemailer,
  (member, identifier) => member.mockImplementation(actual_nodemailer[identifier])
);

const ethereal_timeout_limit = 20000;

const timed_out_flag = "TIMEDOUT";
nodemailer.createTestAccount.mockImplementation( 
  () => promise_timeout_race(
    actual_nodemailer.createTestAccount(),
    ethereal_timeout_limit,
    (resolve, reject) => resolve(timed_out_flag),
  )
);

nodemailer.createTransport.mockImplementation( (transport_config) => {
  // eslint-disable-next-line no-console
  const alert_to_flaked_test = () => console.log(`FLAKY TEST ALERT: was unable to reach ethereal within ${ethereal_timeout_limit}ms, giving up but not failing the test over it.`);

  if ( transport_config.auth === timed_out_flag){
    alert_to_flaked_test();

    return { sendMail: () => ({response: "200"}) };
  }

  const transporter = actual_nodemailer.createTransport(transport_config);

  return {
    ...transporter,
    sendMail: (options) => {
      const send_mail_promise = transporter.sendMail(options);

      const timeout_callback = (resolve, reject) => {
        alert_to_flaked_test();

        resolve({response: "200"});
      };

      return promise_timeout_race(
        send_mail_promise,
        ethereal_timeout_limit,
        timeout_callback,
      );
    },
  };
});


// eslint-disable-next-line no-unused-vars
import { email_backend } from './index.js'; // Server's started as side effect of this import, kind of a GCloud Function thing although I could clean that up


describe("End-to-end tests for email_backend endpoints", () => {
  const prod_test_url = "https://us-central1-report-a-problem-email-244220.cloudfunctions.net/prod-email-backend";
  const local_test_url = `http://127.0.0.1:7331`;
  
  const test_against_prod = false;
  const test_url = test_against_prod ? prod_test_url: local_test_url;

  const make_email_template_names_request = () => axios.get(`${test_url}/email_template_names`);
  const make_email_template_request = (template_name) => axios.get(
    `${test_url}/email_template?template_name=${template_name}`,
    { validateStatus: _.constant(true) } // Don't throw errors on ANY status values, will be intentionally getting some 400's
  );
  const make_submit_email_request = (template_name, completed_template) => axios.post(
    `${test_url}/submit_email`,
    {
      template_name,
      completed_template,
    },
    { validateStatus: _.constant(true) } // Don't throw errors on ANY status values, will be intentionally getting some 400's
  );

  const test_template_name = "test_template.test";
  const completed_test_template = {
    meta: {
      subject_template: "Test subject: [${enums}], ${radio}, ${number}",
    },
  
    enums: ["bug", "other"],
    radio: "yes",
    text: "a",
    number: 1,
    json: {bleh: "bleh", a: 1},
  
    required_automatic: "blah",
    optional_automatic: "bluh",
  };
  
  
  it("/email_template_names returns an array of template names", async () => {
    const { data: template_names } = await make_email_template_names_request();

    const template_names_is_array = _.isArray(template_names);
    const template_names_values_are_strings = _.every(template_names, _.isString);

    return expect(template_names_is_array && template_names_values_are_strings).toBe(true);
  });    


  it("/email_template returns status 400 for an invalid invalid template_name", async () => {
    const { status: bad_template_name_status } = await make_email_template_request("zzz_unlikely_name");

    return expect(bad_template_name_status).toBe(400);
  });

  it("/email_template returns a non-empty object when given a valid template_name", async () => {
    const { data: template } = await make_email_template_request(test_template_name);

    const template_is_valid = template && _.isObject(template) && !_.isEmpty(template);

    return expect(template_is_valid).toBe(true);
  });


  it("/submit_email returns status 400 when a non-existant or invalid template is submitted", async () => {
    const { status: bad_template_name_status } = await make_submit_email_request("zzz_unlikely_name", completed_test_template);
    const { status: invalid_template_status } = await make_submit_email_request(test_template_name, {bleh: "bleh"});

    return expect([bad_template_name_status, invalid_template_status]).toEqual([400, 400]);
  });

  // this test is flaky due to its reliance on a third party service to validate submitted emails; the services is occasionally unresponsive
  it("/submit_email returns status 200 when a valid template is submitted", 
    async () => {
      const { status: ok } = await make_submit_email_request(test_template_name, completed_test_template);

      return expect(ok).toBe(200);
    },
    // timeout on the async returning, just needs to be significantly longer than ethereal_timeout_limit. Shouldn't hit the Jest level timeout as the time-constraint in this test
    // is communications with ethereal, which is being timed out after ethereal_timeout_limit. If an ethereal_timeout_limit is hit, then this test flakes passingly (not great, but it
    // was between that and just dropping this test altogether). Can still flake if the Jest level timeout is hit for nondeterministic system resource/event loop reasons, ah well!
    ethereal_timeout_limit*5
  );
});