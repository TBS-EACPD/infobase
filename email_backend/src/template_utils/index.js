import { get_templates } from './get_templates.js';
import { validate_completed_template } from './validate_completed_template.js';
import { make_email_subject_from_completed_template } from './make_email_subject_from_completed_template.js';
import { make_email_body_from_completed_template } from './make_email_body_from_completed_template.js';

export {
  get_templates,
  validate_completed_template,
  make_email_subject_from_completed_template,
  make_email_body_from_completed_template,
};