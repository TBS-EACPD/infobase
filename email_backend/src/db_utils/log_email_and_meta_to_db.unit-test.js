jest.mock("mongoose"); // eslint-disable-line no-undef
import mongoose from "mongoose";
const actual_mongoose = jest.requireActual("mongoose"); // eslint-disable-line no-undef
mongoose.Schema.mockImplementation(actual_mongoose.Schema);
mongoose.model.mockImplementation((name, schema) => {
  const model = actual_mongoose.model(name, schema);

  model.create = async (record_obj) => {
    let validated = true;
    try {
      await model.validate(record_obj);
    } catch (err) {
      validated = err;
    }
    return validated;
  };

  return model;
});

import { log_email_and_meta_to_db } from "./log_email_and_meta_to_db.js";

describe("log_email_and_meta_to_db", () => {
  const request = { method: "POST", headers: { referer: "http://localhost" } };
  const template_name = "test_template";
  const original_template = {
    meta: {
      subject_template: "Test subject: [${enums}], ${radio}, ${number}",
    },

    enums: {
      required: true,
      value_type: "enums",
      enum_values: {
        bug: {
          en: "Something is broken",
          fr: "Quelque chose ne fonctionne pas",
        },
        typo: {
          en: "It has a spelling or grammar mistake",
          fr: "II y a une erreur d'orthographe ou de grammaire",
        },
        other: {
          en: "Other",
          fr: "Autre",
        },
      },
      form_type: "checkbox",
      form_label: {
        en: "Select all that apply:",
        fr: "Sélectionner toutes les cases qui s'appliquent :",
      },
    },
    radio: {
      required: true,
      value_type: "enums",
      enum_values: {
        yes: {
          en: "Yes",
          fr: "Oui",
        },
        no: {
          en: "No",
          fr: "Non",
        },
      },
      form_type: "radio",
      form_label: {
        en: "Did you find what you were looking for on GC InfoBase?",
        fr: "Avez-vous trouvé ce que vous cherchiez dans InfoBase du GC?",
      },
    },
    text: {
      required: true,
      value_type: "string",
      form_type: "textarea",
      form_label: {
        en: "Text",
        fr: "Text",
      },
    },
    number: {
      required: true,
      value_type: "number",
      form_type: "test",
      form_label: {
        en: "Number",
        fr: "Number",
      },
    },
    json: {
      required: true,
      value_type: "json",
      form_type: "textarea",
      form_label: {
        en: "JSON",
        fr: "JSON",
      },
    },

    required_automatic: {
      required: true,
      value_type: "string",
      form_type: false,
    },
    optional_automatic: {
      required: false,
      value_type: "string",
      form_type: false,
    },
  };
  const completed_template = {
    enums: ["bug", "other"],
    radio: ["yes"],
    text: "a",
    number: 1,
    json: { bleh: "bleh", a: 1 },

    required_automatic: "blah",
    optional_automatic: "bluh",
  };
  const email_config = {
    to: "bleh@blah.ca",
    from: "blah@bleh.com",
  };

  it("Builds a mongoose model from the original template, constructs and saves a valid record based on the completed template", async () => {
    // The model.create call at the end of log_email_and_meta_to_db has been mocked out and replaced with model.validate
    const final_record_object_validated_against_constructed_model =
      await log_email_and_meta_to_db(
        request,
        template_name,
        original_template,
        completed_template,
        email_config
      );

    return expect(
      final_record_object_validated_against_constructed_model
    ).toEqual(true);
  });
});
