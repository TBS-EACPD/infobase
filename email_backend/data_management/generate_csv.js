// const fs = require("fs");
// const _ = require("lodash");
// const json = JSON.parse(
//   fs.readFileSync("./data_management/email_data.json", "utf8")
// );
// const csv_format = [
//   [
//     "ID",
//     "Issue types",
//     "Issue Details",
//     "Commit SHA",
//     "Route",
//     "Language",
//     "App Version",
//     "Client ID",
//     "Recipient",
//     "Sender",
//     "Referer",
//     "Server Time",
//     "Date",
//     "Time",
//   ],
// ];

// const reduced_json = _.map(json, function (report) {
//   return _.reduce(
//     report,
//     function (result, value, key) {
//       if (key === "__v") {
//         return result;
//       }
//       if (_.isPlainObject(value)) {
//         return { ...result, ...value };
//       } else if (_.isArray(value)) {
//         result[key] = value.join(" & ");
//       } else {
//         result[key] = value;
//       }
//       return result;
//     },
//     {}
//   );
// });

// _.map(reduced_json, function (report) {
//   report["server_time"] = new Date(report["server_time"]).toUTCString();
//   csv_format.push(_.values(report));
// });

// const csv = _.map(csv_format, function (format) {
//   return format.join(",");
// });

// fs.writeFile("./data_management/email_data.csv", csv.join("\r\n"), function (
//   err
// ) {
//   console.log(err || "Successfully Saved.");
// });

import { connect_db } from "../transpiled_build/db_utils/connect_db";
import { get_templates } from "../transpiled_build/template_utils";
import { make_mongoose_model_from_original_template } from "../transpiled_build/db_utils/log_email_and_meta_to_db";
import _ from "lodash";
connect_db();
const templates = get_templates();
_.map(templates, (template_value, template_name) => {
  const collection = make_mongoose_model_from_original_template({
    original_template: template_value,
    template_name,
  });

  //if it happens that later on we add another key with a json value,
  //change reduced json to handle "key.value". For the time being, the CSV
  //will be cleaner if we flatten the "key.value" to "value"

  collection.find({}, function (err, submissions) {
    const csv_format = _.map(_.keys(collection.schema.paths), (key) => {
      const key_split = key.split(".");
      return key_split.length > 1 ? key_split[1] : key;
    });
    const csv = [csv_format];

    const reduced_json = _.map(submissions, function (sub) {
      return _.reduce(
        sub._doc,
        function (result, value, key) {
          if (key === "__v") {
            return result;
          }
          if (_.isPlainObject(value)) {
            return { ...result, ...value };
          } else if (_.isArray(value)) {
            result[key] = value.join(" & ");
          } else {
            result[key] = value;
          }
          return result;
        },
        {}
      );
    });

    const time_fixed_json = _.map(reduced_json, function (report) {
      report["server_time"] = new Date(report["server_time"]).toUTCString();
      return report;
    });

    _.map(time_fixed_json, (submission) => {
      csv_format.push(
        _.map(csv_format, (key) => {
          console.log(key);
          return _.has(submission, key) ? submission[key] : "";
        })
      );
    });

    // console.log(csv_format);
  });
});
// const original_template = templates["report_a_problem"];
// const template_name = "report_a_problem";
// const Report = make_mongoose_model_from_original_template({
//   original_template,
//   template_name,
// });

// Report.find({}, function (err, report) {
//   console.log(report);
// });
