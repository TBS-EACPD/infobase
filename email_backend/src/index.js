// entry-point, for both dev and GCF
import { make_email_backend } from "./email_backend.js";
import { get_templates } from "./template_utils";

const email_backend = (() => {
  const templates = get_templates();

  const email_backend = make_email_backend(templates);

  if (!process.env.IS_PROD_SERVER) {
    email_backend.set("port", 7331);
    email_backend.listen(email_backend.get("port"), () => {
      const port = email_backend.get("port");
      //eslint-disable-next-line no-console
      console.log(`InfoBase email backend running at http://127.0.0.1:${port}`);
    });
  }

  return email_backend;
})();

export { email_backend };
