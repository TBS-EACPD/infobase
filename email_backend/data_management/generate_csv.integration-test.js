import _ from "lodash";
import axios from "axios";
import { get_output } from "./generate_csv";
import { connect_db } from "../src/db_utils/connect_db";
import { make_mongoose_model_from_original_template } from "../src/db_utils/log_email_and_meta_to_db";
import { get_templates } from "../src/template_utils";
import fs from "fs";

//Make sure there is test data to work with
const test_template_name = "test_template.test";
const completed_test_template = {
  enums: ["bug", "other"],
  radio: ["yes"],
  text: "a",
  number: 1,
  json: { bleh: "bleh", a: 1 },

  required_automatic: "blah",
  optional_automatic: "bluh",
};

const instance = axios.create({
  headers: {
    referer: "http://localhost:8080/build/InfoBase/index-eng.html",
  },
});
let csv, json, csv_title;

beforeAll((done) => {
  instance
    .post("http://127.0.0.1:7331/submit_email", {
      template_name: test_template_name,
      completed_template: completed_test_template,
    })
    .then(() => {
      get_output().then(({ csv_name, json_output }) => {
        json = json_output;
        fs.readFile(__dirname + `/CSVs/${csv_name}`, "utf-8", (err, data) => {
          if (err) {
            console.log(err);
          } else {
            csv_title = csv_name;
            csv = data;
            done();
          }
        });
      });
    });
});

afterAll((done) => {
  const template = get_templates()[test_template_name];
  connect_db().then(() => {
    make_mongoose_model_from_original_template({
      original_template: template,
      template_name: test_template_name,
    })
      .collection.drop()
      .then(() => {
        fs.unlinkSync(__dirname + `/CSVs/${csv_title}`);
        done();
      });
  });
});

describe("Check that CSV output and JSON output are correct", () => {
  it("Snapshot of JSON output", () => {
    expect(json).toMatchSnapshot({
      _id: expect.any(Object),
      a: 1,
      bleh: "bleh",
      date: expect.any(String),
      enums: "bug, other",
      from: "Sender Name <sender@example.com>",
      number: 1,
      optional_automatic: "bluh",
      radio: "yes",
      referer: "http://localhost:8080/build/InfoBase/index-eng.html",
      required_automatic: "blah",
      server_time: expect.any(String),
      text: "a",
      time: expect.any(String),
      to: "Recipient <recipient@example.com>",
    });
  });

  it("Testing CSV for data", () => {
    expect(csv.split("\n").length).toBe(2);
  });
});
