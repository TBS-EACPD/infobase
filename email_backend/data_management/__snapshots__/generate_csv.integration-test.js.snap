// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check that CSV output and JSON output are correct Snapshot of JSON output 1`] = `
Object {
  "_id": Any<Object>,
  "a": 1,
  "bleh": "bleh",
  "date": Any<String>,
  "enums": "bug, other",
  "from": "Sender Name <sender@example.com>",
  "number": 1,
  "optional_automatic": "bluh",
  "radio": "yes",
  "referer": "http://localhost:8080/build/InfoBase/index-eng.html",
  "required_automatic": "blah",
  "server_time": Any<String>,
  "text": "a",
  "time": Any<String>,
  "to": "Recipient <recipient@example.com>",
}
`;
